File: examples/mobilesam1/python/mobilesam1.py
#!/usr/bin/env python3
# Copyright 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
# SPDX-License-Identifier: Apache-2.0
"""MobileSAM prompt segmentation through the native service."""

import argparse

import cv2
import numpy as np
from spacemit_vision import VisionServiceNative, VisionServiceStatus


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--config", required=True)
    parser.add_argument("--image", help="Default: test_image from YAML")
    parser.add_argument("--box", nargs=4, type=float, default=[190, 70, 460, 280])
    parser.add_argument("--output", default="mobilesam1_result.jpg")
    args = parser.parse_args()
    service = VisionServiceNative.create(args.config)
    image_path = args.image or service.get_default_image()
    if not image_path:
        raise ValueError("Set test_image in YAML or pass --image")
    image = cv2.imread(image_path)
    if image is None:
        raise ValueError("Could not read image")
    x1, y1, x2, y2 = args.box
    status, results = service.infer_image_points(
        image, [[x1, y1], [x2, y2]], [2, 3]
    )
    if status != VisionServiceStatus.OK:
        raise RuntimeError(str(status))
    mask = results[0].mask
    overlay = image.copy()
    overlay[mask != 0] = (30, 144, 144)
    output = cv2.addWeighted(image, 0.5, overlay, 0.5, 0)
    cv2.rectangle(
        output, (round(x1), round(y1)), (round(x2), round(y2)), (0, 255, 0), 2
    )
    if not cv2.imwrite(args.output, output):
        raise RuntimeError("Could not save image")
    print(f"Score: {results[0].score}, mask pixels: {np.count_nonzero(mask)}")
    print(f"Result: {args.output}")


if __name__ == "__main__":
    main()

File: examples/mobilesam1/cpp/mobilesam1.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */
#include <iostream>
#include <string>
#include <stdexcept>

#include <opencv2/opencv.hpp>

#include "vision_service.h"  // NOLINT(build/include_subdir)

int main(int argc, char** argv) {
    const bool help = argc == 2 &&
        (std::string(argv[1]) == "--help" || std::string(argv[1]) == "-h");
    if (help || (argc != 2 && argc != 3 && argc != 4 && argc != 8)) {
        std::cout
            << "Usage: mobilesam1 config [image [output [x1 y1 x2 y2]]]\n"
            << "Defaults: image=test_image from YAML, "
            << "output=mobilesam1_result.jpg, box=190 70 460 280\n";
        return help ? 0 : 1;
    }
    try {
        auto service = VisionService::Create(argv[1], "", true);
        if (!service)
            throw std::runtime_error(VisionService::LastCreateError());
        const std::string image_path =
            argc >= 3 ? argv[2] : service->GetDefaultImage();
        const std::string output_path =
            argc >= 4 ? argv[3] : "mobilesam1_result.jpg";
        if (image_path.empty()) {
            throw std::runtime_error(
                "No input image: set test_image in YAML or pass image");
        }
        VisionServiceRequest request{};
        request.image = cv::imread(image_path);
        if (request.image.empty()) {
            throw std::runtime_error("Could not read image: " + image_path);
        }
        request.point_coords = {{190.0F, 70.0F}, {460.0F, 280.0F}};
        if (argc == 8) {
            request.point_coords = {{std::stof(argv[4]), std::stof(argv[5])},
                                    {std::stof(argv[6]), std::stof(argv[7])}};
        }
        request.point_labels = {2, 3};
        VisionServiceResponse response;
        if (service->Infer(request, &response) != VISION_SERVICE_OK) {
            throw std::runtime_error(service->LastError());
        }
        const auto& mask =
            std::get<vision::Segmentation>(response.results.at(0));
        cv::Mat overlay = request.image.clone();
        overlay.setTo(cv::Scalar(30, 144, 144), *mask.mask);
        cv::Mat rendered;
        cv::addWeighted(request.image, 0.5, overlay, 0.5, 0, rendered);
        cv::rectangle(rendered, request.point_coords[0],
            request.point_coords[1], cv::Scalar(0, 255, 0), 2);
        if (!cv::imwrite(output_path, rendered)) {
            throw std::runtime_error("Could not save result");
        }
        std::cout << "Score: " << mask.score
            << ", mask pixels: " << cv::countNonZero(*mask.mask) << '\n';
        std::cout << "Result: " << output_path << '\n';
    } catch (const std::exception& error) {
        std::cerr << error.what() << '\n';
        return 1;
    }
    return 0;
}
